packages/next/src/server/web/edge-function-wrapper.ts

```diff
@@ -0,0 +1,54 @@
+import type { AdapterOptions } from './adapter'
+import { adapter } from './adapter'
+import type { NextRequest } from './spec-extension/request'
+import type { NextFetchEvent } from './spec-extension/fetch-event'
+import type { NextMiddleware, NextMiddlewareResult } from './types'
+import { RequestAsyncStorageWrapper } from '../async-storage/request-async-storage-wrapper'
+import { requestAsyncStorage } from '../../client/components/request-async-storage.external'
+
+export class EdgeFunctionWrapper {
+  private readonly functionHandler: NextMiddleware
+
+  constructor(handler: NextMiddleware) {
+    this.functionHandler = handler
+  }
+
+  public static wrap(options: AdapterOptions) {
+    // Create the edge function wrapper.
+    const wrapper = new EdgeFunctionWrapper(options.handler)
+
+    // Return the wrapping function.
+    return (opts: AdapterOptions) => {
+      return adapter({
+        ...opts,
+        ...options,
+        // Bind the handler method to the wrapper so it still has context.
+        handler: wrapper.handler.bind(wrapper),
+      })
+    }
+  }
+
+  private async handler(
+    request: NextRequest,
+    evt: NextFetchEvent
+  ): Promise<NextMiddlewareResult> {
+    return RequestAsyncStorageWrapper.wrap(
+      requestAsyncStorage,
+      { req: request },
+      async (requestStore) => {
+        // Get the response from the handler.
+        const response = await this.functionHandler(request, evt)
+
+        evt.waitUntil(
+          Promise.all(
+            requestStore.waitUntil.map((p) =>
+              typeof p === 'function' ? p() : p
```

Comment:
I'm not sure this is correct. The `response` might be ready (e.g. `status` and `headers`) but the body is still streaming, which may add additional `waitUntil` and some of the functions `p` may need to wait for streaming to be done to be called. So it really should be something like:

```
evt.waitUntil(waitForStreamingComplete().then(() => theCurrentCode()))
```

?

Reply:
Agreed with @dvoytenko. Resolving the handler promise doesn't mean that the body finished streaming so we need to actually wait for the end of `response.body` stream to trigger executing scheduled functions on `waitUntil`